### ISSUE TITLE
[java] PreserveStackTrace false positive when using initCause in a separate statement

### ISSUE THREAD

**Issue by jsotuyod** [pmd/pmd#844]
**Affects PMD Version:** 
6.0.0

**Rule:**
[`java-bestpractices/PreserveStackTrace`](https://pmd.github.io/pmd-6.0.0/pmd_rules_java_bestpractices.html#preservestacktrace)

**Description:**
The method [`Throwable.initCause(Throwable cause)`](https://docs.oracle.com/javase/7/docs/api/java/lang/Throwable.html#initCause(java.lang.Throwable)) is not properly understood by PMD as a stack-preserving method unless chained in the  `throw` statement.

**Code Sample demonstrating the issue:**

```java
try {
  // Do something
} catch (final UnsupportedEncodingException e) {
  final ApiException apiEx = new ApiException();
  apiEx.initCause(e);
  throw apiEx;
}
```

**Running PMD through:** Any

**Comment by jsotuyod** [comment]
Notice, if we rewrote the catch as:

```java
} catch (final UnsupportedEncodingException e) {
  throw new ApiException().initCause(e);
}
```

no violation is reported

**Comment by albfernandez** [comment]
I've tried against master and seems to be ok.

```xml
	<test-code>
		<description>#844 PreserveStackTrace false positive when using initCause in a separate statement</description>
		<expected-problems>0</expected-problems>
		<code><![CDATA[
public class Foo {
    public void foo(String a) throws Exception {
        try {
		  // Do something
		} catch (final UnsupportedEncodingException e) {
		  final ApiException apiEx = new ApiException();
		  apiEx.initCause(e);
		  throw apiEx;
		}
    }
}
    ]]></code>
	</test-code>

```

and this other examples are ok too

```xml
	<test-code>
		<description>#844-2 PreserveStackTrace false positive when using initCause in a separate statement</description>
		<expected-problems>0</expected-problems>
		<code><![CDATA[
public class Foo {
    public void foo(String a) throws Exception {
        try {
            int i = Integer.parseInt(a);
        } catch(java.io.FileNotFoundException e) {
        	Exception myException = new Exception();
        	myException.initCause(e);
        	throw myException;
        } catch(java.io.IOException e) {
        	ApiException myException = new ApiException("message", e);
        	myException.doSomething();
        	throw myException;
        } catch(Exception e) {
            throw new Exception().initCause(e);
        }
    }
}
    ]]></code>
	</test-code>
```

**Comment by oowekyala** [comment]
I can't reproduce the bug even when going back to an early December revision

**Comment by jsotuyod** [comment]
I'm closing this, I've found a slight difference between the reported code and the actual code. Thanks everyone
